src/Marten/Services/JsonNetSerializer.cs

```diff
@@ -61,54 +62,50 @@ public void Customize(Action<JsonSerializer> configure)
             _withTypes.TypeNameHandling = TypeNameHandling.Objects;
         }
 
-        public void ToJson(object document, TextWriter writer)
+        public void ToJson(object document, Stream stream)
         {
-            using var jsonWriter = new JsonTextWriter(writer)
+            using var jsonWriter = new JsonTextWriter(new StreamWriter(stream))
             {
                 ArrayPool = _jsonArrayPool,
                 CloseOutput = false,
                 AutoCompleteOnClose = false
             };
 
-
             _serializer.Serialize(jsonWriter, document);
 
-            writer.Flush();
+            jsonWriter.Flush();
         }
 
         public string ToJson(object document)
         {
-            var writer = new StringWriter();
-            ToJson(document, writer);
+            using var stream = new MemoryStream();
+            ToJson(document, stream);
 
-            return writer.ToString();
+            stream.Seek(0, SeekOrigin.Begin);
+            stream.Position = 0;
+            var reader = new StreamReader(stream);
+            return reader.ReadToEnd();
         }
 
         public T FromJson<T>(Stream stream)
         {
-            using var jsonReader = new JsonTextReader(new StreamReader(stream))
+            using var jsonReader = new JsonTextReader(StreamExtensions.GetStreamReader(stream))
             {
                 ArrayPool = _jsonArrayPool,
-                CloseInput = false
+                CloseInput = false,
             };
 
             return _serializer.Deserialize<T>(jsonReader);
         }
 
-        public T FromJson<T>(TextReader reader)
+        public Task<T> FromJsonAsync<T>(Stream stream, CancellationToken cancellationToken = default)
```

Comment:
We need to be using ValueTask.

Reply:
Updated to ValueTask in https://github.com/JasperFx/marten/pull/1685/commits/1ef5e80e9875fa88b5cd902e75eca143937d6488. We should probably review if there are no places in our API that should return ValueTask (however, that's out of scope of this PR).